src/backend/base/langflow/services/telemetry/opentelemetry.py

```diff
@@ -1,30 +1,219 @@
+from typing import Dict, Mapping, Tuple, Union
 
+import threading
+
+# a default OpenTelelmetry meter name
+langflow_meter_name = "langflow"
+
+"""
+If the measurement values are non-additive, use an Asynchronous Gauge.
+    ObservableGauge reports the current absolute value when observed.
+If the measurement values are additive: If the value is monotonically increasing - use an Asynchronous Counter.
+If the value is NOT monotonically increasing - use an Asynchronous UpDownCounter.
+    UpDownCounter reports changes/deltas to the last observed value.
+If the measurement values are additive and you want to observe the distribution of the values - use a Histogram.
+"""
+
+
+class MetricType(Enum):
+    COUNTER = "counter"
+    OBSERVABLE_GAUGE = "observable_gauge"
+    HISTOGRAM = "histogram"
+    UP_DOWN_COUNTER = "up_down_counter"
+
+
+class ObservableGaugeWrapper:
+    """
+    Wrapper class for ObservableGauge
+    Since OpenTelemetry does not provide a way to set the value of an ObservableGauge,
+    instead it uses a callback function to get the value, we need to create a wrapper class.
+    """
+
+    def __init__(self, name: str, description: str, unit: str):
+        self._values: Dict[Tuple[Tuple[str, str], ...], float] = {}
+        self._meter = metrics.get_meter(langflow_meter_name)
+        self._gauge = self._meter.create_observable_gauge(
+            name=name, description=description, unit=unit, callbacks=[self._callback]
+        )
+
+    def _callback(self, options: CallbackOptions):
+        return [Observation(value, attributes=dict(labels)) for labels, value in self._values.items()]
+
+        # return [Observation(self._value)]
+
+    def set_value(self, value: float, labels: Mapping[str, str]):
+        self._values[tuple(sorted(labels.items()))] = value
+
+
+class Metric:
+    def __init__(
+        self,
+        name: str,
+        description: str,
+        type: MetricType,
+        unit: str = "",
+    ):
+        self.name = name
+        self.description = description
+        self.type = type
+        self.unit = unit
+
+    def __repr__(self):
+        return f"Metric(name='{self.name}', description='{self.description}', type={self.type}, unit='{self.unit}')"
+
+
+class ThreadSafeSingletonMeta(type):
+    """
+    Thread-safe Singleton metaclass
+    """
+
+    _instances = None
+    _lock: threading.Lock = threading.Lock()
+
+    def __call__(cls, *args, **kwargs):
+        if cls._instances is None:
+            with cls._lock:
+                if cls._instances is None:
+                    cls._instances = super(ThreadSafeSingletonMeta, cls).__call__(*args, **kwargs)
+
+        return cls._instances
+
+
+class OpenTelemetry(metaclass=ThreadSafeSingletonMeta):
+    """
+    Define any custom metrics here
+    A thread safe singleton class to manage metrics
+    """
+
+    _metrics_registry: Mapping[str, Metric] = dict[str, Metric](
+        {
+            "file_uploads": Metric(
```

Comment:
I wonder if we can remove the declaration repetition of the name. maybe having a method that creates a Metric object and call it once per metric to register

e.g.
```
_metrics_registry: Mapping[str, Metric] = dict[str, Metric]


def _add_metric(name: str, description: str, type...):
  self._metrics_registry[name] = ...

self._add_metric("file_uploads"....)
```

Reply:
done